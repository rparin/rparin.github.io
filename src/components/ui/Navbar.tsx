import { NavItems } from "@/constants/NavLinks";

export default function Navbar() {
  const NavTags = [];
  for (const [key, value] of Object.entries(NavItems)) {
    NavTags.push(
      <li key={value.id}>
        <a
          className="text-xl"
          href={value?.url ? value?.url : "#" + value.id}
          target={value?.url ? "_blank" : undefined}
          rel={value?.url ? "noopener noreferrer" : undefined}>
          {value.title}
        </a>
      </li>
    );
  }

  return (
    <>
      <nav className="sticky top-0 bg-background drop-shadow-lg z-10 ">
        <ul className="flex flex-row gap-9 md:gap-12 md:justify-center py-5 ml-5 pr-5 overflow-y-auto">
          {NavTags}
        </ul>
      </nav>
    </>
  );
}
