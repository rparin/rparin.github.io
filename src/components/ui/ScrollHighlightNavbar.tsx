"use client";

import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";

/**
 * @param {number} currentPosition Current Scroll position
 * @param {Array} sectionPositionArray Array of positions of all sections
 * @param {number} startIndex Start index of array
 * @param {number} endIndex End index of array
 * @return {number} Current Active index
 */
const nearestIndex = (
  currentPosition: any,
  sectionPositionArray: any,
  startIndex: any,
  endIndex: any
): any => {
  if (startIndex === endIndex) return startIndex;
  else if (
    Math.ceil(window.innerHeight + window.scrollY) >=
    document.documentElement.scrollHeight
  ) {
    return endIndex;
  } else if (startIndex === endIndex - 1) {
    if (
      Math.abs(
        sectionPositionArray[startIndex].headerRef.current.offsetTop -
          currentPosition
      ) <
      Math.abs(
        sectionPositionArray[endIndex].headerRef.current.offsetTop -
          currentPosition
      )
    )
      return startIndex;
    else return endIndex;
  } else {
    var nextNearest = ~~((startIndex + endIndex) / 2);
    var a = Math.abs(
      sectionPositionArray[nextNearest].headerRef.current.offsetTop -
        currentPosition
    );
    var b = Math.abs(
      sectionPositionArray[nextNearest + 1].headerRef.current.offsetTop -
        currentPosition
    );
    if (a < b) {
      return nearestIndex(
        currentPosition,
        sectionPositionArray,
        startIndex,
        nextNearest
      );
    } else {
      return nearestIndex(
        currentPosition,
        sectionPositionArray,
        nextNearest,
        endIndex
      );
    }
  }
};

export default function ScrollHighlightNavbar({
  navHeader,
}: {
  navHeader: any;
}) {
  const [activeIndex, setActiveIndex] = useState(0);
  useEffect(() => {
    const handleScroll = (e: any) => {
      var index = nearestIndex(
        window.scrollY,
        navHeader,
        0,
        navHeader.length - 1
      );
      setActiveIndex(index);
    };
    document.addEventListener("scroll", handleScroll);
    return () => {
      document.removeEventListener("scroll", handleScroll);
    };
  }, []);

  return (
    <div className="sticky top-0 bg-background drop-shadow-lg z-10">
      {navHeader.map((header: any, index: any) => (
        <a
          key={index + header.headerID}
          style={{ backgroundColor: activeIndex === index ? "red" : "green" }}
          href={`#${header.headerID}`}>
          {header.headerTitle}
        </a>
      ))}
    </div>
  );
}

ScrollHighlightNavbar.propTypes = {
  navHeader: PropTypes.arrayOf(
    PropTypes.shape({
      headerID: PropTypes.string,
      headerRef: PropTypes.object.isRequired,
      headerTitle: PropTypes.string.isRequired,
    })
  ).isRequired,
};
