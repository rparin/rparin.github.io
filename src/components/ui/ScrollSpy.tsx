"use client";
import React, { MutableRefObject, useRef, useState, useEffect } from "react";

export default function ScrollSpy({
  navbar,
  children,
}: {
  navbar: React.ReactElement;
  children: React.ReactElement[];
}) {
  const navIndex: any = {};
  const [activeIndex, setActiveIndex] = useState<number>(0);
  const sections: React.JSX.Element[] = [];
  const navRefs = useRef<Array<HTMLDivElement>>([]);

  const addToRefs = (el: HTMLDivElement) => {
    if (el && !navRefs.current.includes(el)) {
      navRefs.current.push(el);
    }
  };

  children.map((child, index) => {
    if (index == 0) {
      navIndex[index] = "";
      sections.push(<div key="__" ref={addToRefs}></div>);
    }

    navIndex[index + 1] = child.props.id;
    sections.push(
      <child.type
        {...child.props}
        key={index + child.props.id}
        ref={addToRefs}
      />
    );
  });

  useEffect(() => {
    const handleScroll = (e: Event) => {
      var index = nearestIndex(
        window.scrollY,
        navRefs.current,
        0,
        navRefs.current.length - 1
      );
      setActiveIndex(index);
    };

    document.addEventListener("scroll", handleScroll);
    return () => {
      document.removeEventListener("scroll", handleScroll);
    };
  }, []);

  const nv = React.cloneElement(navbar, { activeId: navIndex[activeIndex] });

  return (
    <>
      {nv}
      {sections}
    </>
  );
}

/**
 * @param {number} currentPosition Current Scroll position
 * @param {Array} sectionPositionArray Array of positions of all sections
 * @param {number} startIndex Start index of array
 * @param {number} endIndex End index of array
 * @return {number} Current Active index
 */
const nearestIndex = (
  currentPosition: number,
  sectionPositionArray: Array<HTMLDivElement>,
  startIndex: number,
  endIndex: number
): number => {
  if (startIndex === endIndex) return startIndex;
  else if (
    Math.ceil(window.innerHeight + window.scrollY) >=
    document.documentElement.scrollHeight
  ) {
    return endIndex;
  } else if (startIndex === endIndex - 1) {
    if (
      Math.abs(sectionPositionArray[startIndex].offsetTop - currentPosition) <
      Math.abs(sectionPositionArray[endIndex].offsetTop - currentPosition)
    )
      return startIndex;
    else return endIndex;
  } else {
    var nextNearest = ~~((startIndex + endIndex) / 2);
    var a = Math.abs(
      sectionPositionArray[nextNearest].offsetTop - currentPosition
    );
    var b = Math.abs(
      sectionPositionArray[nextNearest + 1].offsetTop - currentPosition
    );
    if (a < b) {
      return nearestIndex(
        currentPosition,
        sectionPositionArray,
        startIndex,
        nextNearest
      );
    } else {
      return nearestIndex(
        currentPosition,
        sectionPositionArray,
        nextNearest,
        endIndex
      );
    }
  }
};
