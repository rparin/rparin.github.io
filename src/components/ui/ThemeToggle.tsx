"use client";

import { Moon, Sun } from "lucide-react";
import { useTheme } from "next-themes";
import { useState } from "react";
import { cn } from "@/lib/utils";

export function ThemeToggle({ className }: { className?: string | undefined }) {
  const { setTheme, theme } = useTheme();
  const [checked, setCheck] = useState(false);

  const changeTheme = () => {
    setTheme(theme == "dark" ? "light" : "dark");
    setCheck(!checked);
  };

  return (
    <button
      type="button"
      role="switch"
      aria-checked={checked}
      aria-label={`Toggle Theme`}
      className={cn(
        "border-input hover:text-accent-foreground h-10 border bg-background px-4 hover:bg-accent",
        className
      )}
      onClick={() => changeTheme()}>
      <Sun className="dark:hidden" />
      <Moon className="hidden dark:block" />
    </button>
  );
}
