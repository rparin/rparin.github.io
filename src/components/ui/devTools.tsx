"use client";

import Image from "next/image";
import { pType } from "@/constants/Projects";
import { useRef, useState, useEffect } from "react";

export default function DevTools({ tools }: { tools: pType["icons"] }) {
  const [activeTool, setActiveTool] = useState(-1);
  const devRefs = useRef<Array<HTMLAnchorElement>>([]);
  const addToRefs = (el: HTMLAnchorElement) => {
    if (el && !devRefs.current.includes(el)) {
      devRefs.current.push(el);
    }
  };

  const icon_cn = "h-auto w-7 md:w-10 lg:w-14 md:p-1";
  const devTools = tools.map((icon) => {
    return (
      <li key={icon.alt}>
        <a
          aria-label={icon.name}
          ref={addToRefs}
          tabIndex={-1}
          href={icon.url}
          target={icon.url ? "_blank" : undefined}
          rel={icon.url ? "noopener noreferrer" : undefined}>
          <Image
            aria-hidden="true"
            className={icon.invert ? icon_cn + "invert-0 dark:invert" : icon_cn}
            src={icon.img}
            alt={icon.alt}
          />
        </a>
      </li>
    );
  });

  useEffect(() => {
    if (activeTool != -1 && devRefs.current[activeTool]) {
      devRefs.current[activeTool].focus();
    }
  }, [activeTool]);

  const keyListNav = (e: React.KeyboardEvent<HTMLUListElement>) => {
    switch (e.key) {
      case "ArrowDown":
        e.preventDefault();
        if (activeTool === tools.length - 1) {
          setActiveTool(0);
        } else {
          setActiveTool(activeTool + 1);
        }
        break;

      case "ArrowUp":
        e.preventDefault();
        if (activeTool === 0) {
          setActiveTool(tools.length - 1);
        } else {
          setActiveTool(activeTool - 1);
        }
        break;

      case "Home":
        e.preventDefault();
        setActiveTool(0);
        break;

      case "End":
        e.preventDefault();
        setActiveTool(tools.length - 1);
        break;

      default:
        //Nothing
        break;
    }
  };

  return (
    <ul
      aria-label="Dev Tools"
      className="p-scroll row-span-3 mx-1 mt-1 flex flex-col items-center justify-start gap-2 overflow-y-auto md:row-span-4 lg:mt-2"
      onKeyDown={keyListNav}>
      {devTools}
    </ul>
  );
}
