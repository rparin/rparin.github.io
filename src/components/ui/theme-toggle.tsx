"use client";

import { Moon, Sun } from "lucide-react";
import { useTheme } from "next-themes";
import { Button } from "@/components/ui/button";
import { useEffect } from "react";
import { signal } from "@preact/signals";

export const signalTheme = signal("light");
export function ThemeToggle({ className }: { className?: string | undefined }) {
  const { setTheme } = useTheme();

  const setThemeHelper = (theme: "dark" | "light" | string) => {
    setTheme(theme);
    signalTheme.value = theme;
  };

  const changeTheme = () => {
    const savedTheme = window.localStorage.getItem("theme");
    setThemeHelper(savedTheme == "dark" ? "light" : "dark");
  };

  useEffect(() => {
    const savedTheme = window.localStorage.getItem("theme");
    if (savedTheme) {
      setThemeHelper(savedTheme);
    } else {
      const isDark = window.matchMedia("(prefers-color-scheme: dark)").matches;
      setThemeHelper(isDark ? "dark" : "light");
    }
  });

  return (
    <Button
      className={className}
      variant="outline"
      onClick={() => changeTheme()}>
      <Sun className="h-[2rem] w-[1.2rem] rotate-0 scale-100 transition-all dark:-rotate-90 dark:scale-0" />
      <Moon className="absolute h-[2rem] w-[1.2rem] rotate-90 scale-0 transition-all dark:rotate-0 dark:scale-100" />
      <span className="sr-only">Toggle theme</span>
    </Button>
  );
}
