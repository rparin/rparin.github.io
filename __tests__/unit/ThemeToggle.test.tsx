import { render } from "@testing-library/react";
import { axe, toHaveNoViolations } from "jest-axe";
import { ThemeToggle } from "@/components/ui/ThemeToggle";

describe("testing ThemeToggle Component Accessibility", () => {
  expect.extend(toHaveNoViolations);
  test("it should have no accessibility violations", async () => {
    const { container } = render(<ThemeToggle />);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

test("it renders component unchanged", () => {
  const { container } = render(<ThemeToggle />);
  expect(container).toMatchSnapshot();
});
