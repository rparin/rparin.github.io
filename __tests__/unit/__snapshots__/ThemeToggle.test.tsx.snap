// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders component unchanged 1`] = `
<div>
  <button
    aria-checked="false"
    aria-label="Toggle Theme"
    class="border-input hover:text-accent-foreground h-10 border bg-background px-4 hover:bg-accent"
    role="switch"
    type="button"
  >
    <svg
      class="dark:hidden"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="4"
      />
      <path
        d="M12 2v2"
      />
      <path
        d="M12 20v2"
      />
      <path
        d="m4.93 4.93 1.41 1.41"
      />
      <path
        d="m17.66 17.66 1.41 1.41"
      />
      <path
        d="M2 12h2"
      />
      <path
        d="M20 12h2"
      />
      <path
        d="m6.34 17.66-1.41 1.41"
      />
      <path
        d="m19.07 4.93-1.41 1.41"
      />
    </svg>
    <svg
      class="hidden dark:block"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 3a6 6 0 0 0 9 9 9 9 0 1 1-9-9Z"
      />
    </svg>
  </button>
</div>
`;
